hail/python/hailtop/aiotools/fs.py

```diff
@@ -458,45 +553,53 @@ async def _dest_type(self, transfer):
 
         return dest_type
 
-    async def copy_source(self, transfer, src, dest_type_task):
+    async def copy_source(self, worker_pool: AsyncWorkerPool, transfer: Transfer, source_report: SourceReport, src, dest_type_task):
         src_copier = SourceCopier(self.router_fs, src, transfer.dest, transfer.treat_dest_as, dest_type_task)
-        await src_copier.copy()
-
-    async def _copy_one_transfer(self, transfer: Transfer):
-        dest_type_task = asyncio.create_task(self._dest_type(transfer))
-        dest_type_task_awaited = False
+        await src_copier.copy(worker_pool, source_report)
 
+    async def _copy_one_transfer(self, worker_pool: AsyncWorkerPool, transfer_report: TransferReport, transfer: Transfer):
         try:
-            src = transfer.src
-            if isinstance(src, list):
-                if transfer.treat_dest_as == Transfer.TARGET_FILE:
-                    raise NotADirectoryError(transfer.dest)
-
-                for s in src:
-                    await self.copy_source(transfer, s, dest_type_task)
-            else:
-                await self.copy_source(transfer, src, dest_type_task)
-
-            # raise potential exception
-            dest_type_task_awaited = True
-            await dest_type_task
-        finally:
-            if not dest_type_task_awaited:
-                # retrieve dest_type_task exception to avoid
-                # "Task exception was never retrieved" errors
-                try:
-                    dest_type_task_awaited = True
-                    await dest_type_task
-                except:
-                    pass
+            dest_type_task = asyncio.create_task(self._dest_type(transfer))
+            dest_type_task_awaited = False
 
-    async def copy(self, transfer: Union[Transfer, List[Transfer]]):
-        if isinstance(transfer, Transfer):
-            await self._copy_one_transfer(transfer)
-            return
+            try:
+                src = transfer.src
+                if isinstance(src, str):
+                    await self.copy_source(worker_pool, transfer, transfer_report._source_report, src, dest_type_task)
+                else:
+                    if transfer.treat_dest_as == Transfer.TARGET_FILE:
+                        raise NotADirectoryError(transfer.dest)
+
+                    async with WaitableSharedPool(worker_pool) as pool:
+                        for r, s in zip(transfer_report._source_report, src):
```

Comment:
Did you intend to continue trying to copy all files even if one fails? Opposite behavior of gsutil?

Reply:
Yes.  This is designed for maximum parallelism.  So by the time one source fails, all the other sources in the transfer have potentially (and likely) already started.  I see two alternatives:
 - Copy things serially.  Obviously we don't like that, we're designing for performance.
 - Cancel the partial copies when one fails.  I don't like that much either, and would rather one fail and one succeed than one fail and one get stopped part way through.